src/framework/smoke_queue.hpp

```diff
@@ -0,0 +1,161 @@
+#ifndef SERENITY_SMOKE_QUEUE_HPP
+#define SERENITY_SMOKE_QUEUE_HPP
+
+#include <memory>
+#include <string>
+#include <random>
+#include <vector>
+
+#include <mesos/resources.hpp>
+
+#include <stout/json.hpp>
+#include <stout/option.hpp>
+
+#include "logging/logging.hpp"
+
+#include "smoke_job.hpp"
+
+struct QueueBlock {
+  QueueBlock(std::shared_ptr<SmokeJob> _job,
+              double_t _probability,
+              size_t _index)
+    : baseJob(_job),
+      originalProbability(_probability),
+      baseProbability(_probability),
+      index(_index),
+      aliasJob(nullptr),
+      ready(false) {}
+
+  double_t originalProbability;
+  std::shared_ptr<SmokeJob> baseJob;
+  double_t baseProbability;
+
+  std::shared_ptr<SmokeJob> aliasJob;
+
+  size_t index;
+  bool ready;
+
+  void print() {
+    LOG(INFO) << "{ id: " << baseJob->id
+      << ", baseProbability: " << baseProbability
+      << ", aliasId: "
+      << (aliasJob != nullptr ? std::to_string(aliasJob->id) : "<none>")
+      << "}";
+  }
+};
+
+
+class SmokeAliasQueue {
+ public:
+  SmokeAliasQueue() : totalShares(0), initialized(false),
+                 localIndex(0), finished(false) {}
+
+  /**
+   * Using Naive Alias Method from http://www.keithschwarz.com/darts-dice-coins/
+   */
+  void init() {
+    totalShares = 0;
+    for (auto& job : this->jobs) {
+      // Reset default values.
+      job.baseProbability = job.originalProbability;
+      totalShares += job.baseJob->shares;
+      job.ready = false;
+    }
+
+
+    double_t blockShares = totalShares / (double_t) this->jobs.size();
+    intRand = std::uniform_int_distribution<uint32_t>(0, this->jobs.size() - 1);
+    doubleRand = std::uniform_real_distribution<double_t>(0, blockShares);
+
+    for (int i =1; i < this->jobs.size(); i++) {
+      QueueBlock* underfullBlock = nullptr;
+      QueueBlock* overfullBlock = nullptr;
+
+      for (auto& job : this->jobs) {
+        if (job.ready) continue;
+        if (job.baseProbability >= blockShares) {
+          overfullBlock = &job;
+          break;
+        }
+      }
+
+      BOOST_ASSERT(overfullBlock != nullptr);
```

Comment:
Are you sure those asserts work in our builds?
Take in mind that they won't fire in Release build (if they really work like casserts as documentation states).

If you want to make sure that your pointer is not null all the time, it's more portable to use 'if () return;'